distmv/normal.go

```diff
@@ -118,3 +121,109 @@ func (n *Normal) Rand(x []float64) []float64 {
+// n - len(dims), where n is the dimension of the original receiver. The updated
+// mean and covariance are
+//  mu = mu_1 + sigma_{2,1}^T * sigma_{2,2}^-1 (v - mu_2)
+//  sigma = sigma_{1,1} - sigma_{2,1}^T * sigma_{2,2}^-1 * sigma_{2,1}
+// where the elements of x_2 have been observed with values v. The dimension order
+// is preserved during conditioning, so if the value of dimension 1 is observed,
+// the returned normal represents dimensions {0, 2, ...} of the original Normal
+// distribution.
+//
+// ConditionalNormal returns {nil, false} if there is a failure during the update.
+// Mathematically this is impossible, but can occur with finite precision arithmetic.
+func (n *Normal) ConditionNormal(observed []int, values []float64, src *rand.Rand) (*Normal, bool) {
+	if len(observed) != len(values) {
+		panic("normal: input slice length mismatch")
+	}
+
+	sigma := n.getSigma()
+
+	ob := len(observed)
+	unob := n.Dim() - ob
+	obMap := make(map[int]struct{})
+	for _, v := range observed {
+		if _, ok := obMap[v]; ok {
+			panic("normal: observed dimension occurs twice")
+		}
+		obMap[v] = struct{}{}
+	}
+	unobserved := make([]int, 0, unob)
+	for i := 0; i < n.Dim(); i++ {
+		if _, ok := obMap[i]; !ok {
+			unobserved = append(unobserved, i)
+		}
+	}
+	mu1 := make([]float64, unob)
+	for i, v := range unobserved {
+		mu1[i] = n.mu[v]
+	}
+	mu2 := make([]float64, ob) // really v - mu2
+	for i, v := range observed {
+		mu2[i] = values[i] - n.mu[v]
+	}
+
+	var sigma11, sigma22 mat64.SymDense
+	sigma11.SubsetSym(sigma, unobserved)
+	sigma22.SubsetSym(sigma, observed)
+
+	sigma21 := mat64.NewDense(ob, unob, nil)
+	for i, r := range observed {
+		for j, c := range unobserved {
+			v := sigma.At(r, c)
+			sigma21.Set(i, j, v)
+		}
+	}
+
+	var chol mat64.Cholesky
+	ok := chol.Factorize(&sigma22)
+	if !ok {
+		return nil, ok
+	}
+
+	// Compute sigma_{2,1}^T * sigma_{2,2}^-1 (v - mu_2).
+	v := mat64.NewVector(ob, mu2)
+	var tmp, tmp2 mat64.Vector
+	err := tmp.SolveCholeskyVec(&chol, v)
+	if err != nil {
+		return nil, false
+	}
+	tmp2.MulVec(sigma21.T(), &tmp)
+
+	// Compute sigma_{2,1}^T * sigma_{2,2}^-1 * sigma_{2,1}.
+	// TODO(btracey): Should this be a method of SymDense?
+	var tmp3, tmp4 mat64.Dense
+	err = tmp3.SolveCholesky(&chol, sigma21)
+	if err != nil {
+		return nil, false
+	}
+	tmp4.Mul(sigma21.T(), &tmp3)
+
+	for i := range mu1 {
+		mu1[i] += tmp2.At(i, 0)
+	}
+
+	// TODO(btracey): If tmp2 can constructed with a method, then this can be
+	// replaced with SubSym.
+	for i := 0; i < len(unobserved); i++ {
+		for j := i; j < len(unobserved); j++ {
+			v := sigma11.At(i, j)
+			sigma11.SetSym(i, j, v-tmp4.At(i, j))
+		}
+	}
+	return NewNormal(mu1, &sigma11, src)
+}
+
+// sigma computes and stores the covariance matrix of the distribution, or
+// returns it if it has already been computed.
+func (n *Normal) getSigma() *mat64.SymDense {
+	n.once.Do(func() {
+		sigma := mat64.NewSymDense(n.Dim(), nil)
```

Comment:
> Do is intended for initialization that must be run exactly once. Since f is niladic, it may be necessary to use a function literal to capture the arguments to a function to be invoked by Do:

```
config.once.Do(func() { config.init(filename) })
```

>Because *no call to Do returns until the one call to f returns*, if f causes Do to be called, it will deadlock.

This means that if you have reached a use of n.sigma later, it must be non-nil. If this weren't true, your return approach would not work either.